### PR TITLE
Fix #4813: [java] SwitchStmtsShouldHaveDefault false positive with pattern matching

### LINKED ISSUE #4813: [java] SwitchStmtsShouldHaveDefault false positive with pattern matching
**Affects PMD Version:** 7.0.0-rc4 

**Rule:**

Please provide the rule name and a link to the rule documentation:
- https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#switchstmtsshouldhavedefault
- https://docs.pmd-code.org/latest/pmd_rules_java_performance.html#toofewbranchesforaswitchstatement

**Description:**

**Code Sample demonstrating the issue:**

```java
    public sealed interface AcceptableResult permits Success, Warning {
        public String message();
    }

    public final class Success implements AcceptableResult {

        @Override
        public String message() {
            return "Success!";
        }
    }

    public abstract class Failure {

        abstract public String message();
    }

    public non-sealed class Warning extends Failure implements AcceptableResult {

        @Override
        public String message() {
            return "Oopsie";
        }
    }

    public class ProviderWarning extends Warning {

        @Override
        public String message() {
            return "Ohoh";
        }
    }

    public void test(AcceptableResult result) {
        switch (result) {
            case Warning failure -> System.out.println("WARNING " + failure.message());
            case Success success -> System.out.println(success.message());
        }
    }
```

**Expected outcome:**

PMD reports violations on the switch SwitchStmtsShouldHaveDefault and TooFewBranchesForASwitchStatement. For SwitchStmtsShouldHaveDefault it's a false positive since we have covered all of the cases here. As for TooFewBranchesForASwitchStatement I think it's also a false positive in this case, since using a pattern matching switch is the cleanest way to do this kind of stuff even if there are only 2 branches.  

**Running PMD through:** * Maven*

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

Switch Statements, that use pattern matching are simply ignored.

## Related issues

- Fixes #4813

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-java/src/main/resources/category/java/bestpractices.xml [hunk 1]
```diff
@@ -1168,12 +1168,18 @@ class SomeTestClass {
...
+            compiler already ensures that all cases are covered. The same is true for Switch Expressions,
+            which are also not considered by this rule.
         </description>
         <priority>3</priority>
         <properties>
             <property name="xpath">
                 <value><![CDATA[
                 //SwitchStatement[@DefaultCase = false() and @ExhaustiveEnumSwitch = false()]
+                    (: exclude type pattern tests - for these, the compiler will ensure exhaustiveness :)
+                    [not(*/SwitchLabel/TypePattern)]
```
Comment (adangel) [inline pmd-java/src/main/resources/category/java/bestpractices.xml:1182]: Should be probably something like

```suggestion
                    [not(*/SwitchLabel/(TypePattern|RecordPattern))]
```

And a separate test case for record patterns is needed?

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to main:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 4 violations, 2 errors and 7 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/11183047084?pr=5252">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/resources/category/java/bestpractices.xml [hunk 2]
```diff
@@ -1168,12 +1168,18 @@ class SomeTestClass {
...
+            compiler already ensures that all cases are covered. The same is true for Switch Expressions,
+            which are also not considered by this rule.
         </description>
         <priority>3</priority>
         <properties>
             <property name="xpath">
                 <value><![CDATA[
                 //SwitchStatement[@DefaultCase = false() and @ExhaustiveEnumSwitch = false()]
+                    (: exclude pattern tests - for these, the compiler will ensure exhaustiveness :)
+                    [not(*/SwitchLabel/(TypePattern|RecordPattern))]
```
Comment (adangel) [inline pmd-java/src/main/resources/category/java/bestpractices.xml:1182]: After #5247 is merged:

```suggestion
                    (: exclude pattern tests - for these, the compiler will ensure exhaustiveness :)
                    [not(*/SwitchLabel[@PatternLabel = true()])]
```

Comment (adangel) [reply pmd-java/src/main/resources/category/java/bestpractices.xml:1182]: Or alternatively:

```
//SwitchStatement
    [@EnumSwitch = true()]
    [@DefaultCase = false()]
    [@ExhaustiveEnumSwitch = false()]
```

Comment (adangel) [reply pmd-java/src/main/resources/category/java/bestpractices.xml:1182]: Not _that_ simple - we should also report a violation for a switch over a int without a default - which is not an enum...

```suggestion
                    (: exclude pattern tests - for these, the compiler will ensure exhaustiveness :)
                    [*/SwitchLabel[@PatternLabel = false()]]
```

**Pull request merged**